lib/Sema/CSSimplify.cpp

```diff
@@ -1358,10 +1359,10 @@ static bool matchFunctionRepresentations(FunctionTypeRepresentation rep1,
   case ConstraintKind::BindParam:
   case ConstraintKind::BindToPointerType:
   case ConstraintKind::Equal:
-    return rep1 != rep2;
+  case ConstraintKind::Subtype:
```

Comment:
I think the reason why it's done like that is because there is a subtyping rule between some of the representations which is verified/maintained by SIL, is that right @slavapestov? I guess if we wanted to change status quo here we might add actual checks whether there is a subtyping rule between given two function representations and if not - add a tailored fix.

Reply:
Humm, interesting ... I thought that maybe could be a rule of something like this, but I did a test on this example changing `c` convention to all others and `thin, block, swift` and except when the representation was the same all crashed, so my guess was that it only allowed this way... I don't understand much about SIL or how it verifies this but let me try to take a look
Thanks @xedin :)

Reply:
To be honest I don't know that the subtype matrix here is e.g.

```
func foo(_: (@convention(c) () -> Void) -> Void) {}
func bar(_: () -> Void) {}

foo(bar)
```

It's accepted by type-checker but crashes with assert in SILGen. Looks like CSApply generates a conversion between `() -> Void` and `@conversion(c) () -> Void`. Maybe we should reject this code in Sema. @jckarter, thoughts?

Reply:
Seems like the same happens with `block` convention too, `thin` doesn't. 
For what I could see on SIL(although I don't understand much) is that SIL representations thin -> thick are conversions are allowed, but only in some situations ...

Reply:
Right, I think I'm having progress, but question... if a function `foo`  type has as parameter a `@convetion(c)` function type which makes its representation `c`, the representation of the function `foo` should still be swift?
@xedin @slavapestov @jckarter

Reply:
I think that's correct, arguments do not affect representation of the declaration.

Reply:
Right, I change the approach to take into account if even though the outter function representation is swift, if it has a foreign function as param, `emitCFunctionPointer`. This is wip because still only fixes the situation @xedin mentioned where the destination is foreign.
```swift
func foo(_: (@convention(c) () -> Void) -> Void) {}
func bar(_: () -> Void) {}

foo(bar)// now diagnose correctly 
```
I still need to figure out the case where the source is the c/block ... but is good to get a feedback already :)

cc @slavapestov

Reply:
> I think that's correct, arguments do not affect representation of the declaration.

That makes sense @xedin, but for this, we need to take into account the representation of params in order to emit the correct c function on SILGen. As I mentioned in the previous comment I made some progress let's see if we can figure this out :)

Reply:
We normally allow conversions between different conversions at the type system level, but there is a post-type-checking pass that checks conversions from Swift or `block` convention to `@convention(c)` and raises an error if the value is not a global function reference. It looks like we fail to catch cases where the convention change occurs in the argument of an outer function type.

Reply:
> It looks like we fail to catch cases where the convention change occurs in the argument of an outer function type.

@jckarter As far as I could see, different kinds of conversions in the argument context hit different crashes. You mention `@convention(c)` should be convertible for other conversions this should be also applicable in the argument context? Should be this allowed at a validation level and fixed to be supported?

Reply:
@jckarter Just to clarify what you are saying is that we should forbid any conversions between different representations when function types are embedded into other type e.g. optional, argument/result of a function, generic arguments?

Reply:
@xedin If it works today, I would still allow it for conversions from "thin" conventions (c, thin) to "thick" (swift, block) or conversions between thick conventions. It's only thick-to-thin conversions that can't be supported (except for the limited cases where we can tell that they statically reference a global function that doesn't require context).

Reply:
Thank you for clarification! I think we can use locator in `matchFunctionTypes` to find where function type is inner and reject certain conversions in Sema.

Reply:
Ok, I'll get back to this later today after work :) 
Just to clarify, sorry I don't know if I fully understand the "except for the limited cases where we can tell that they statically reference a global function that doesn't require context" but from what I get here is we are forbidding different representations checking for only argument locator function types?

Reply:
@xedin I've made some changes, let me know what you think :)